chainstate/tx-verifier/src/transaction_verifier/input_check/mod.rs

```diff
@@ -29,7 +31,58 @@ use mintscript::{
 
 use crate::TransactionVerifierStorageRef;
 
-use super::{error::ConnectTransactionError, TransactionSourceForConnect};
+use super::TransactionSourceForConnect;
+
+pub type TimelockError = mintscript::checker::TimelockError<TimelockContextError>;
+pub type ScriptError = mintscript::script::ScriptError<DestinationSigError, TimelockError>;
+
+#[derive(PartialEq, Eq, Clone, thiserror::Error, Debug)]
+pub enum InputCheckErrorPayload {
+    #[error("Utxo {0:?} missing or spent")]
+    MissingUtxo(common::chain::UtxoOutPoint),
+
+    #[error("Utxo view error: {0}")]
+    UtxoView(#[from] utxo::Error),
+
+    #[error(transparent)]
+    Translation(#[from] mintscript::translate::TranslationError),
+
+    #[error(transparent)]
+    Verification(#[from] ScriptError),
+}
+
+impl From<mintscript::script::ScriptError<Infallible, TimelockError>> for InputCheckErrorPayload {
+    fn from(value: mintscript::script::ScriptError<Infallible, TimelockError>) -> Self {
+        Self::Verification(value.errs_into())
+    }
+}
+
+#[derive(PartialEq, Eq, Clone, thiserror::Error, Debug)]
+#[error("Error verifying input #{input_num}: {error}")]
+pub struct InputCheckError {
+    input_num: usize,
```

Comment:
input number alone won't be that usefull if block contains several txs. Is it easy to add source_id?

Reply:
> input number alone won't be that usefull if block contains several txs. Is it easy to add source_id?

I'd say that should be done at different level in the transaction verifier by technique similar to here. I.e. blck verification procedure attaches the block ID to the error, transaction verification attaches transaction ID to all transaction verification errors.

The full error would then look something like:
```
Error verifying block 0xaabbcc: error verifying transaction 0xb1ab1a: Error in input #4: Bad signature or whatever
```

Reply:
ok agree